File: clients/go/debug_interface.go
package main

import (
    "bufio"
    . "ai_cup_22/model"
    . "ai_cup_22/codegame"
    . "ai_cup_22/debugging"
)

type DebugInterface struct {
    Reader *bufio.Reader
    Writer *bufio.Writer
}

func (debugInterface DebugInterface) AddPlacedText(position Vec2, text string, alignment Vec2, size float64, color Color) {
    debugInterface.Add(NewDebugDataPlacedText(position, text, alignment, size, color))
}

func (debugInterface DebugInterface) AddCircle(position Vec2, radius float64, color Color) {
    debugInterface.Add(NewDebugDataCircle(position, radius, color))
}

func (debugInterface DebugInterface) AddGradientCircle(position Vec2, radius float64, innerColor Color, outerColor Color) {
    debugInterface.Add(NewDebugDataGradientCircle(position, radius, innerColor, outerColor))
}

func (debugInterface DebugInterface) AddRing(position Vec2, radius float64, width float64, color Color) {
    debugInterface.Add(NewDebugDataRing(position, radius, width, color))
}

func (debugInterface DebugInterface) AddPie(position Vec2, radius float64, startAngle float64, endAngle float64, color Color) {
    debugInterface.Add(NewDebugDataPie(position, radius, startAngle, endAngle, color))
}

func (debugInterface DebugInterface) AddArc(position Vec2, radius float64, width float64, startAngle float64, endAngle float64, color Color) {
    debugInterface.Add(NewDebugDataArc(position, radius, width, startAngle, endAngle, color))
}

func (debugInterface DebugInterface) AddRect(bottomLeft Vec2, size Vec2, color Color) {
    debugInterface.Add(NewDebugDataRect(bottomLeft, size, color))
}

func (debugInterface DebugInterface) AddPolygon(vertices []Vec2, color Color) {
    debugInterface.Add(NewDebugDataPolygon(vertices, color))
}

func (debugInterface DebugInterface) AddGradientPolygon(vertices []ColoredVertex) {
    debugInterface.Add(NewDebugDataGradientPolygon(vertices))
}

func (debugInterface DebugInterface) AddSegment(firstEnd Vec2, secondEnd Vec2, width float64, color Color) {
    debugInterface.Add(NewDebugDataSegment(firstEnd, secondEnd, width, color))
}

func (debugInterface DebugInterface) AddGradientSegment(firstEnd Vec2, firstColor Color, secondEnd Vec2, secondColor Color, width float64) {
    debugInterface.Add(NewDebugDataGradientSegment(firstEnd, firstColor, secondEnd, secondColor, width))
}

func (debugInterface DebugInterface) AddPolyLine(vertices []Vec2, width float64, color Color) {
    debugInterface.Add(NewDebugDataPolyLine(vertices, width, color))
}

func (debugInterface DebugInterface) AddGradientPolyLine(vertices []ColoredVertex, width float64) {
    debugInterface.Add(NewDebugDataGradientPolyLine(vertices, width))
}

func (debugInterface DebugInterface) Add(debugData DebugData) {
    debugInterface.Send(NewDebugCommandAdd(debugData))
}

func (debugInterface DebugInterface) Clear() {
    debugInterface.Send(NewDebugCommandClear())
}

func (debugInterface DebugInterface) SetAutoFlush(enable bool) {
    debugInterface.Send(NewDebugCommandSetAutoFlush(enable))
}

func (debugInterface DebugInterface) Flush() {
    debugInterface.Send(NewDebugCommandFlush())
}

func (debugInterface DebugInterface) Send(command DebugCommand) {
    ClientMessageDebugMessage{
        Command: command,
    }.Write(debugInterface.Writer)
    err := debugInterface.Writer.Flush()
    if err != nil {
        panic(err)
    }
}

func (debugInterface DebugInterface) GetState() DebugState {
    ClientMessageRequestDebugState{}.Write(debugInterface.Writer)
    err := debugInterface.Writer.Flush()
    if err != nil {
        panic(err)
    }
    return ReadDebugState(debugInterface.Reader)
}

File: clients/fsharp/DebugInterface.fs
namespace AiCup22

type DebugInterface(reader, writer) =
    member this.addPlacedText(position: Model.Vec2, text: string, alignment: Model.Vec2, size: double, color: Debugging.Color) =
        this.add (Debugging.DebugData.PlacedText {
            Position = position
            Text = text
            Alignment = alignment
            Size = size
            Color = color
        })
    
    member this.addCircle(position: Model.Vec2, radius: double, color: Debugging.Color) =
        this.add (Debugging.DebugData.Circle {
            Position = position
            Radius = radius
            Color = color
        })
    
    member this.addGradientCircle(position: Model.Vec2, radius: double, innerColor: Debugging.Color, outerColor: Debugging.Color) =
        this.add (Debugging.DebugData.GradientCircle {
            Position = position
            Radius = radius
            InnerColor = innerColor
            OuterColor = outerColor
        })
    
    member this.addRing(position: Model.Vec2, radius: double, width: double, color: Debugging.Color) =
        this.add (Debugging.DebugData.Ring {
            Position = position
            Radius = radius
            Width = width
            Color = color
        })
    
    member this.addPie(position: Model.Vec2, radius: double, startAngle: double, endAngle: double, color: Debugging.Color) =
        this.add (Debugging.DebugData.Pie {
            Position = position
            Radius = radius
            StartAngle = startAngle
            EndAngle = endAngle
            Color = color
        })
    
    member this.addArc(position: Model.Vec2, radius: double, width: double, startAngle: double, endAngle: double, color: Debugging.Color) =
        this.add (Debugging.DebugData.Arc {
            Position = position
            Radius = radius
            Width = width
            StartAngle = startAngle
            EndAngle = endAngle
            Color = color
        })
    
    member this.addRect(bottomLeft: Model.Vec2, size: Model.Vec2, color: Debugging.Color) =
        this.add (Debugging.DebugData.Rect {
            BottomLeft = bottomLeft
            Size = size
            Color = color
        })
    
    member this.addPolygon(vertices: Model.Vec2[], color: Debugging.Color) =
        this.add (Debugging.DebugData.Polygon {
            Vertices = vertices
            Color = color
        })
    
    member this.addGradientPolygon(vertices: Debugging.ColoredVertex[]) =
        this.add (Debugging.DebugData.GradientPolygon {
            Vertices = vertices
        })
    
    member this.addSegment(firstEnd: Model.Vec2, secondEnd: Model.Vec2, width: double, color: Debugging.Color) =
        this.add (Debugging.DebugData.Segment {
            FirstEnd = firstEnd
            SecondEnd = secondEnd
            Width = width
            Color = color
        })
    
    member this.addGradientSegment(firstEnd: Model.Vec2, firstColor: Debugging.Color, secondEnd: Model.Vec2, secondColor: Debugging.Color, width: double) =
        this.add (Debugging.DebugData.GradientSegment {
            FirstEnd = firstEnd
            FirstColor = firstColor
            SecondEnd = secondEnd
            SecondColor = secondColor
            Width = width
        })
    
    member this.addPolyLine(vertices: Model.Vec2[], width: double, color: Debugging.Color) =
        this.add (Debugging.DebugData.PolyLine {
            Vertices = vertices
            Width = width
            Color = color
        })
    
    member this.addGradientPolyLine(vertices: Debugging.ColoredVertex[], width: double) =
        this.add (Debugging.DebugData.GradientPolyLine {
            Vertices = vertices
            Width = width
        })

    member this.add(debugData: Debugging.DebugData) =
        this.send (Debugging.DebugCommand.Add {
            DebugData = debugData
        })
    
    member this.clear() =
        this.send (Debugging.DebugCommand.Clear (new Debugging.DebugCommandClear()))
    
    member this.setAutoFlush(enable: bool) =
        this.send (Debugging.DebugCommand.SetAutoFlush {
            Enable = enable
        })
    
    member this.flush() =
        this.send (Debugging.DebugCommand.Flush (new Debugging.DebugCommandFlush()))

    member this.send(command: Debugging.DebugCommand) =
        (Codegame.ClientMessage.DebugMessage { Command = command }).writeTo writer
        writer.Flush()

    member this.getState(): Debugging.DebugState =
        (new Codegame.ClientMessageRequestDebugState()).writeTo writer
        writer.Flush()
        Debugging.DebugState.readFrom reader